bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/WebSocketV2.java

```diff
@@ -70,79 +124,198 @@ public void onWebSocketText(@Nullable String msgarg) {
+            remoteControllerWebSocket.callback.currentAppUpdated(jsonMsg.getName());
+        }
+        if (currentSourceApp.equals(jsonMsg.getId()) && "false".equals(jsonMsg.getVisible())) {
+            currentSourceApp = "";
+            remoteControllerWebSocket.callback.currentAppUpdated("");
         }
     }
 
     @NonNullByDefault({})
-    static class JSONAppStatus {
-        public JSONAppStatus(String id) {
+    class JSONApp {
+        public JSONApp(String id, String method) {
+            this(id, method, null);
+        }
+
+        public JSONApp(String id, String method, @Nullable String metaTag) {
+            // use message id to identify app to remove
             this.id = id;
+            this.method = method;
             params.id = id;
+            // not working
+            params.metaTag = metaTag;
         }
 
-        @NonNullByDefault({})
-        static class Params {
+        class Params {
             String id;
+            String metaTag;
         }
 
-        String method = "ms.application.get";
+        String method;
         String id;
         Params params = new Params();
     }
 
+    /**
+     * update manApp.name if it's incorrect
+     */
+    void updateApps(JSONMessage jsonMsg) {
+        remoteControllerWebSocket.manApps.values().stream()
+                .filter(a -> a.getAppId().equals(jsonMsg.getId()) && !a.getName().equals(jsonMsg.getName()))
+                .peek(a -> logger.trace("{}: Updated app name {} to: {}", host, a.getName(), jsonMsg.getName()))
+                .findFirst().ifPresent(a -> a.setName(jsonMsg.getName()));
+
+        updateApp(jsonMsg);
+    }
+
+    /**
+     * Fix app key, if it's the app id
+     */
+    @SuppressWarnings("null")
+    void updateApp(JSONMessage jsonMsg) {
+        if (remoteControllerWebSocket.manApps.containsKey(jsonMsg.getId())) {
+            int type = remoteControllerWebSocket.manApps.get(jsonMsg.getId()).getType();
+            remoteControllerWebSocket.manApps.put(jsonMsg.getName(),
+                    remoteControllerWebSocket.new App(jsonMsg.getId(), jsonMsg.getName(), type));
+            remoteControllerWebSocket.manApps.remove(jsonMsg.getId());
+            logger.trace("{}: Updated app id {} name to: {}", host, jsonMsg.getId(), jsonMsg.getName());
+            remoteControllerWebSocket.updateCount = 0;
+        }
+    }
+
+    /**
+     * Send get application status
+     *
+     * @param id appId of app to get status for
+     */
     void getAppStatus(String id) {
-        sendCommand(remoteControllerWebSocket.gson.toJson(new JSONAppStatus(id)));
+        if (!id.isEmpty()) {
+            boolean appType = getAppStream().filter(a -> a.getAppId().equals(id)).map(a -> a.getType() == 2).findFirst()
+                    .orElse(true);
+            // note apptype 4 always seems to return an error, so use default of 2 (true)
+            String apptype = (appType) ? "ms.application.get" : "ms.webapplication.get";
```

Comment:
These are not the same strings, and this is the only place they are used.

Reply:
Exactly, because these strings are used on multiple places, it is better to use a class level constant that can be re-used on all places. This would make the code more maintainable.